bindings/pyroot/pythonizations/test/sofie_gnn.py

```diff
@@ -0,0 +1,317 @@
+    r = input_data["receivers"]
+    return {
+      "globals" : np.zeros((GLOBAL_FEATURE_SIZE )),
+      "nodes"   : np.zeros((n.shape[0],NODE_FEATURE_SIZE )),
+      "edges"   : np.zeros((e.shape[0],EDGE_FEATURE_SIZE )),
+      "senders" : s, "receivers" : r
+    }
+
+LATENT_SIZE = 2
+def make_mlp_model():
+  """Instantiates a new MLP, followed by LayerNorm.
+
+  The parameters of each new MLP are not shared with others generated by
+  this function.
+
+  Returns:
+    A Sonnet module which contains the MLP and LayerNorm.
+  """
+  return snt.Sequential([
+      snt.nets.MLP([LATENT_SIZE]*2, activate_final=True),
+      snt.LayerNorm(axis=-1, create_offset=True, create_scale=True)
+  ])
+
+def CopyData(input_data) :
+  output_data = ROOT.TMVA.Experimental.SOFIE.Copy(input_data)
+  return output_data
+
+class MLPGraphIndependent(snt.Module):
+  """GraphIndependent with MLP edge, node, and global models."""
+
+  def __init__(self, name="MLPGraphIndependent"):
+    super(MLPGraphIndependent, self).__init__(name=name)
+    self._network = gn.modules.GraphIndependent(
+        edge_model_fn = lambda: snt.nets.MLP([LATENT_SIZE]*2, activate_final=True),
+        node_model_fn = lambda: snt.nets.MLP([LATENT_SIZE]*2, activate_final=True),
+        global_model_fn = lambda: snt.nets.MLP([LATENT_SIZE]*2, activate_final=True))
+
+  def __call__(self, inputs):
+    return self._network(inputs)
+
+
+class MLPGraphNetwork(snt.Module):
+  """GraphNetwork with MLP edge, node, and global models."""
+
+  def __init__(self, name="MLPGraphNetwork"):
+    super(MLPGraphNetwork, self).__init__(name=name)
+    self._network = gn.modules.GraphNetwork(
+            edge_model_fn=make_mlp_model,
+            node_model_fn=make_mlp_model,
+            global_model_fn=make_mlp_model)
+
+  def __call__(self, inputs):
+    return self._network(inputs)
+
+def PrintGData(data, printShape = True):
+  n = data.nodes.numpy()
+  e = data.edges.numpy()
+  g = data.globals.numpy()
+  if (printShape) :
+    print("GNet data ... shapes",n.shape,e.shape,g.shape)
+  print(" node data", n.reshape(n.size,))
+  print(" edge data", e.reshape(e.size,))
+  print(" global data",g.reshape(g.size,))
+
+class EncodeProcessDecode(snt.Module):
+
+  def __init__(self,
+               name="EncodeProcessDecode"):
+    super(EncodeProcessDecode, self).__init__(name=name)
+    self._encoder = MLPGraphIndependent()
+    self._core = MLPGraphNetwork()
+    self._decoder = MLPGraphIndependent()
+    self._output_transform = MLPGraphIndependent()
+
+  def __call__(self, input_op, num_processing_steps):
+    latent = self._encoder(input_op)
+    latent0 = latent
+    output_ops = []
+    for _ in range(num_processing_steps):
+      core_input = utils_tf.concat([latent0, latent], axis=1)
+      latent = self._core(core_input)
+      decoded_op = self._decoder(latent)
+      output_ops.append(self._output_transform(decoded_op))
+    return output_ops
+
+class SOFIE_GNN(unittest.TestCase):
```

Comment:
It appears there is an issue with destroying the headers generated by SOFIE before the end of the Python process responsible for running the tests in this class. While this issue may be investigated in a separate PR, I believe it is important we make sure the new tests created in this PR do not increase the failures in our CI. My suggestion is to write a `tearDownClass` method of this class, that removes all the headers created by the various tests in this class, only once at the end.

```python
class SOFIE_GNN(unittest.TestCase):
    def tearDownClass(cls):
        for name in cls.generated_headers:
            os.remove(name)
            
```

This means that tests need to keep track of the filenames being created and store them in a class variable

Reply:
@lmoneta made the change.